### PR TITLE
chore(release-pr): update `upload-artifact` action to v4

### PR DESCRIPTION
**Pull request by Juneezee**
## Change Summary

v3 of `actions/upload-artifact` and `actions/download-artifact` will be fully deprecated by **30 January 2025**. Jobs that are scheduled to run during the brownout periods will also fail. See:

1. https://github.blog/changelog/2024-04-16-deprecation-notice-v3-of-the-artifact-actions/
2. https://github.blog/changelog/2024-11-05-notice-of-breaking-changes-for-github-actions/
3. https://github.blog/changelog/2024-12-05-notice-of-upcoming-releases-and-breaking-changes-for-github-actions/#artifacts-v3-brownouts

## Change type

- [ ] feat: (new feature for the user, not a new feature for build script)
- [ ] fix: (bug fix for the user, not a fix to a build script)
- [ ] docs: (changes to the documentation)
- [ ] style: (formatting, missing semi colons, etc; no production code change)
- [ ] refactor: (refactoring production code, eg. renaming a variable)
- [ ] test: (adding missing tests, refactoring tests; no production code change)
- [x] chore: (updating grunt tasks etc; no production code change)

## Test/ Verification

Provide summary of changes.

## Additional information / screenshots (optional)

Anything for maintainers to be made aware of

### CODE REVIEW

**Comment by Juneezee** [comment]
@dstala @mertmit friendly ping for review :blush:

**Review by dstala — APPROVED** [review]

**Review by mertmit — CHANGES_REQUESTED** [review]
Thank you for the contribution!

It seems we are using a pattern which is not supported on v4 (in v4 artifacts are immutable)
We need to migrate in a proper way otherwise actions will be broken.

ref: https://github.com/actions/upload-artifact/blob/main/docs/MIGRATION.md

**Comment by Juneezee** [comment]
@mertmit The `preview-spec` name is used for all three jobs. I can update them to have unique names now (e.g. `preview-spec-pr`, `preview-spec-uffizzi`), or do you have any specific naming preferences?

File: .github/workflows/release-pr.yml [hunk 1]
```diff
@@ -174,8 +174,9 @@ jobs:
...
           EOF
       - name: Upload PR Event as Artifact
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
-          name: preview-spec
+          name: preview-spec-event
           path: event.json
           retention-days: 2
+          overwrite: true
```
Comment (Juneezee) [inline .github/workflows/release-pr.yml:182]: This follows the `v3` behaviour by overwriting the `event.json` if the PR is closed.

https://github.com/actions/upload-artifact/blob/main/docs/MIGRATION.md#overwriting-an-artifact

File: .github/workflows/uffizzi-preview.yml [hunk 2]
```diff
@@ -20,31 +20,14 @@ jobs:
...
-               owner: context.repo.owner,
-               repo: context.repo.repo,
-               artifact_id: matchArtifact.id,
-               archive_format: 'zip',
-            });
-            let fs = require('fs');
-            fs.writeFileSync(`${process.env.GITHUB_WORKSPACE}/preview-spec.zip`, Buffer.from(download.data));
+          path: preview-spec
+          pattern: preview-spec-*
+          merge-multiple: true
```
Comment (Juneezee) [inline .github/workflows/uffizzi-preview.yml:27]: We can simplify this step by using `actions/download-artifact@v4`. It supports `pattern` and `merge-multiple` so all artifacts uploaded by `release-pr.yml` can be downloaded into the `preview-spec` directory

https://github.com/actions/download-artifact?tab=readme-ov-file#download-multiple-filtered-artifacts-to-the-same-directory

**Review by Juneezee — COMMENTED** [review]
Explaining my code changes

**Review by mertmit — APPROVED** [review]

**Pull request merged**